DuckDuckGo/LocalNotificationsLogic.swift

```diff
@@ -0,0 +1,256 @@
+            if let status = store.scheduleStatus(for: notification),
+                case let ScheduleStatus.scheduled(date) = status,
+                date < currentDate {
+                store.didFire(notification: notification)
+                
+                let pixel: PixelName
+                switch notification {
+                case .privacy:
+                    pixel = .notificationD0Fired
+                case .homeRow:
+                    pixel = .notificationD1Fired
+                }
+                
+                LocalNotifications.shared.checkPermissions { (authorization, alertSettings) in
+                    let parameters = [Keys.authorization: String(authorization.rawValue),
+                                      Keys.alert: String(alertSettings.rawValue)]
+                    Pixel.fire(pixel: pixel, withAdditionalParameters: parameters)
+                }
+            }
+        }
+    }
+    
+    private func cancelPrivacyNotification() {
+        LocalNotifications.shared.cancelNotifications(withIdentifiers: [Notification.privacy.identifier])
+        store.didCancel(notification: .privacy)
+    }
+    
+    func willLeaveApplication() {
+        
+        if variantManager.isSupported(feature: .dayZeroNotification),
+            store.notificationsEnabled,
+            store.scheduleStatus(for: .privacy) == nil {
+            schedulePrivacyNotification()
+        }
+        
+        if variantManager.isSupported(feature: .dayOneNotification),
+            store.notificationsEnabled,
+            store.scheduleStatus(for: .homeRow) == nil {
+            scheduleHomeRowNotification()
+        }
+    }
+        
+    private func schedulePrivacyNotification() {
+        let privacyData = PrivacyReportDataSource()
+        
+        let title: String
+        let body: String
+        if privacyData.trackersCount < Constants.privacyNotificationInfoTreshold {
+            title = "Ready to take back your privacy?"
+            body = "Keep browsing with DuckDuckGo to block trackers and encrypt connections."
+        } else {
+            title = "Success! The internet just got less creepy."
+            body = "While using DuckDuckGo you blocked \(privacyData.trackersCount) trackers and secured \(privacyData.httpsUpgradesCount) unencrypted connections."
+        }
+        
+        LocalNotifications.shared.scheduleNotification(title: title,
+                                                       body: body,
+                                                       identifier: Notification.privacy.identifier,
+                                                       timeInterval: Constants.privacyNotificationDelay)
+        store.didSchedule(notification: .privacy, date: Date().addingTimeInterval(Constants.privacyNotificationDelay))
+    }
+    
+    func fireDateForHomeRowNotification(currentDate: Date = Date()) -> (DateComponents, Date)? {
```

Comment:
This tuple seems to be redundant?

Reply:
Yes, leftover from the time when I wanted to use UICalendar trigger (that requires components) but encountered some bugs that were difficult to resolve. I'll clean this up.